### PR TITLE
Enforce single context per term invariant

### PR DESCRIPTION
**Pull request by oflatt**

### CODE REVIEW

File: dag_in_context/src/interval_analysis.egg [hunk 1]
```diff
@@ -269,39 +269,6 @@
       ((set (hi-bound lhs) (bound-max hi-thn hi-els)))
       :ruleset interval-analysis)
 
-; If the If takes a single value
-; If input < v is true, then upper bound of input is upper bound of v - 1
-(rule (
```
Comment (oflatt) [inline dag_in_context/src/interval_analysis.egg:274]: Unnecessary rules- all our if statements have tuple inputs

File: dag_in_context/src/utility/subst.egg [hunk 2]
```diff
@@ -31,12 +31,6 @@
        (subsume (Subst fuel1 assum to in)))
       :ruleset subst-helpers)
 
-;; Base case- leaf nodes
```
Comment (oflatt) [inline dag_in_context/src/utility/subst.egg:34]: Got rid of case without context

File: dag_in_context/src/utility/subst.rs [hunk 3]
```diff
@@ -197,88 +210,3 @@ fn test_subst_preserves_context() -> crate::Result {
         vec![],
     )
 }
-
```
Comment (oflatt) [inline dag_in_context/src/utility/subst.rs:200]: substitution doesn't need fuel any more- cleanup PR coming soon

File: dag_in_context/src/utility/context_of.egg [hunk 4]
```diff
@@ -6,14 +6,13 @@
...
 ; Error checking - each expr should only have a single context
-; TODO: this may not be true
-; (rule ((ContextOf x ctx1)
-;         (ContextOf x ctx2)
-;         (!= ctx1 ctx2))
-;     (
-;         (panic "Equivalent expressions have nonequivalent context")
-;     )
-;     :ruleset error-checking)
+(rule ((ContextOf x ctx1)
```
Comment (oflatt) [inline dag_in_context/src/utility/context_of.egg:9]: We now enforce this invariant

File: dag_in_context/src/utility/context-prop.egg [hunk 5]
```diff
@@ -1,58 +1,39 @@
...
-(rule ((InLoop pred body))
-      ((ScopeContext (InLoop pred body)))
-      :ruleset context-prop)
-(rule ((NoContext))
-      ((ScopeContext (NoContext)))
-      :ruleset context-prop)
 
 (rewrite (InContext ctx (Top op a b c))
          (Top op (InContext ctx a) (InContext ctx b) (InContext ctx c))
-         :when ((ScopeContext ctx))
```
Comment (oflatt) [inline dag_in_context/src/utility/context-prop.egg:20]: scopecontext isn't a thing any more- all contexts are scope contexts

File: dag_in_context/src/type_analysis.egg [hunk 6]
```diff
@@ -114,6 +114,8 @@
        (!= ty ty2))
       ((panic "if branches else branch has incorrect input type"))
       :ruleset type-analysis)
+
```
Comment (ajpal) [inline dag_in_context/src/type_analysis.egg:117]: Why?

**Review by ajpal — APPROVED** [review]

File: dag_in_context/src/type_analysis.egg (hunk 6 above)
Comment (oflatt) [reply dag_in_context/src/type_analysis.egg:117]: no reason

**Pull request merged**